### PR TITLE
core: Add OFI_BUFPOOL_MMAP flag to bufpool

### PR DESCRIPTION
**Pull request by nzmsv**
Add OFI_BUFPOOL_MMAP flag that requests that the buffer pool allocations are satisfied by mmap() rather than malloc().

This is required to safely use MADV_DONTFORK when registering the buffers with ibv_reg_mr in the efa provider.

This is a refinement of the change originally submitted as https://github.com/ofiwg/libfabric/pull/7431 and is required to resolve a memory corruption in the efa provider.

### CODE REVIEW

**Review by wzamazon — CHANGES_REQUESTED** [review]

File: src/unix/osd.c [hunk 1]
```diff
@@ -120,6 +120,26 @@ int ofi_wait_cond(pthread_cond_t *cond, pthread_mutex_t *mut, int timeout_ms)
 	return pthread_cond_timedwait(cond, mut, &ts);
 }
 
+int ofi_mmap_anon_pages(void **memptr, size_t size)
+{
+	size = ofi_get_aligned_size(size, ofi_get_page_size());
+	*memptr = mmap(NULL, size, PROT_READ | PROT_WRITE,
```
[the remark this answers is not in the document]
Comment (shefty) [reply src/unix/osd.c:126]: If addr is NULL, then mmap will return a page-aligned mapping.

File: prov/util/src/util_buf.c [hunk 2]
```diff
@@ -45,6 +45,44 @@ enum {
...
+			buf_region->flags = OFI_BUFPOOL_HUGEPAGES;
+			return 0;
+		}
+		/* If we can't allocate huge pages, fall back to normal
+		 * allocations for all future attempts.
+		 */
+		pool->attr.flags &= ~OFI_BUFPOOL_HUGEPAGES;
+	}
+
+	if (pool->attr.flags & OFI_BUFPOOL_MMAP) {
```
Comment (shefty) [inline prov/util/src/util_buf.c:66]: Other locations use else if after checking for huge pages, which is what I think you want here as well.  hugepages and mmap are exclusive from each other.

File: src/unix/osd.c [hunk 3]
```diff
@@ -120,6 +120,26 @@ int ofi_wait_cond(pthread_cond_t *cond, pthread_mutex_t *mut, int timeout_ms)
...
+		MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+	if (OFI_UNLIKELY(*memptr == MAP_FAILED)) {
+		return -ofi_syserr();
```
Comment (rashikakheria) [inline src/unix/osd.c:129]: Shouldn't this be just `errno`?

File: include/windows/osd.h [hunk 4]
```diff
@@ -682,6 +682,15 @@ static inline char* strsep(char **stringp, const char *delim)
 
 #define __attribute__(x)
 
+static inline int ofi_mmap_anon_pages(void **memptr, size_t size)
+{
+	return -FI_ENOSYS;
+}
+
+static inline void ofi_unmap_anon_pages(void *memptr, size_t size)
+{
```
Comment (rashikakheria) [inline include/windows/osd.h:691]: Why isn't this returning `-FI_ENOSYS`?

File: src/unix/osd.c [hunk 5]
```diff
@@ -120,6 +120,26 @@ int ofi_wait_cond(pthread_cond_t *cond, pthread_mutex_t *mut, int timeout_ms)
 	return pthread_cond_timedwait(cond, mut, &ts);
 }
 
+int ofi_mmap_anon_pages(void **memptr, size_t size)
+{
+	size = ofi_get_aligned_size(size, ofi_get_page_size());
```
Comment (rashikakheria) [inline src/unix/osd.c:125]: Should we return the corrected size to caller?

File: prov/util/src/util_buf.c (hunk 2 above)
Comment (rashikakheria) [inline prov/util/src/util_buf.c:66]: Shouldn't this be an else if? I don't see a case where we will try mmap after trying allocating hugepages...

File: src/unix/osd.c (hunk 3 above)
Comment (nzmsv) [reply src/unix/osd.c:129]: That's what the wrapper returns.  Seeing as this is platform-specific code, sure, I can replace this with a direct use of `errno`.

File: include/windows/osd.h (hunk 4 above)
Comment (nzmsv) [reply include/windows/osd.h:691]: I don't think there is anything we can do on failure to unmap a buffer, and we should never call this because we'll never succeed in allocating a buffer with the MMAP flag set.

File: prov/util/src/util_buf.c (hunk 2 above)
Comment (nzmsv) [reply prov/util/src/util_buf.c:66]: This is on purpose.  I originally wrote this to fall through to mmap if huge page allocation failed, but decided to leave this to follow-up discussion.

The only time we'll ask for huge pages is when fork support is not needed, so in theory it would be safe to just use small pages as before.  However, it may be nice to move the buffers involved with `ibv_reg_mr` into a separate mmap anyway.  This could be accomplished by keeping the fall-through here and specifying `OFI_BUFPOOL_HUGEPAGES | OFI_BUFPOOL_MMAP` in efa (see https://github.com/nzmsv/libfabric/pull/1)

Comment (nzmsv) [reply prov/util/src/util_buf.c:66]: This is on purpose.  I originally wrote this to fall through to mmap if huge page allocation failed, but decided to leave this to follow-up discussion.

The only time efa asks for huge pages is when fork support is not needed, so in theory it would be safe to just use small pages as before.  However, it may be nice to move the buffers involved with `ibv_reg_mr` into a separate mmap anyway.  This could be accomplished by keeping the fall-through here and specifying `OFI_BUFPOOL_HUGEPAGES | OFI_BUFPOOL_MMAP` in efa (see https://github.com/nzmsv/libfabric/pull/1).

Thoughts?

Comment (rashikakheria) [reply prov/util/src/util_buf.c:66]: Since, its a core change (not just specific to EFA provider), there may be cases where providers would allocate from hugepages or fallback to malloc. I think if provider would like an option to try mmap after failed hugepage allocation then they can use enable both the flags as you suggested.

Comment (shefty) [reply prov/util/src/util_buf.c:66]: Hmm... huge page allocation is via mmap.  So I'm thinking that the definition should be:
```
OFI_BUFPOOL_MMAP = 1 << 3,
OFI_BUFPOOL_HUGEPAGES = (1 << 4) | OFI_BUFPOOL_MMAP,
```
The fallback from huge pages mmap -> malloc could be considered an error since we're no longer guaranteed to return page aligned allocations.  (Even falling back from mmap to malloc is questionable in that regard, but so unlikely to happen that I think we're okay.)

Comment (shefty) [reply prov/util/src/util_buf.c:66]: Basically, I think your original choice to fallback to mmap is what we want.

Comment (nzmsv) [reply prov/util/src/util_buf.c:74]: @shefty also agreed (and provided the same feedback) so I am keeping this

File: src/unix/osd.c (hunk 5 above)
Comment (nzmsv) [reply src/unix/osd.c]: New revision calculates the size outside of this function.

File: include/windows/osd.h (hunk 4 above)
Comment (nzmsv) [reply include/windows/osd.h:691]: Changed to return errno to unify this with hugepage unmap.

File: prov/util/src/util_buf.c (hunk 2 above)
Comment (nzmsv) [reply prov/util/src/util_buf.c:74]: I decided to keep the MMAP and HUGEPAGE flags separate because we may want to unset HUGEPAGE individually (for example, when it fails, or when the allocation is smaller than the size of a single hugepage) but keep the mmap behavior.

File: prov/util/src/util_buf.c [hunk 6]
```diff
@@ -149,6 +171,8 @@ int ofi_bufpool_grow(struct ofi_bufpool *pool)
 err2:
 	if (buf_region->flags & OFI_BUFPOOL_HUGEPAGES)
 		ofi_free_hugepage_buf(buf_region->alloc_region, pool->alloc_size);
+	else if (buf_region->flags & OFI_BUFPOOL_MMAP)
```
[the remark this answers is not in the document]
Comment (nzmsv) [reply prov/util/src/util_buf.c]: Done!

File: prov/util/src/util_buf.c [hunk 7]
```diff
@@ -45,6 +45,44 @@ enum {
 };
 
 
+static int ofi_bufpool_alloc(struct ofi_bufpool_region *buf_region)
```
[the remark this answers is not in the document]
Comment (nzmsv) [reply prov/util/src/util_buf.c]: Done!

File: include/unix/osd.h [hunk 8]
```diff
@@ -90,6 +90,9 @@ struct util_shm
 	size_t		size;
 };
 
+int ofi_mmap_anon_pages(void **memptr, size_t size);
+void ofi_unmap_anon_pages(void *memptr, size_t size);
```
[the remark this answers is not in the document]
Comment (nzmsv) [reply include/unix/osd.h]: I combined these

File: include/ofi_mem.h [hunk 9]
```diff
@@ -289,6 +289,7 @@ enum {
 	OFI_BUFPOOL_INDEXED		= 1 << 1,
 	OFI_BUFPOOL_NO_TRACK		= 1 << 2,
 	OFI_BUFPOOL_HUGEPAGES		= 1 << 3,
+	OFI_BUFPOOL_MMAP		= 1 << 4,
```
Comment (shefty) [inline include/ofi_mem.h:292]: I'm saying that HUGEPAGES implies MMAP.  That's how huge pages are allocated.  So, these aren't separate.  It's the reverse that's not true (MMAP -/-> HUGEPAGES).  That's why I would define HUGEPAGES = MMAP | BIT(x).

Then the code path below would automatically fall back from huge pages to mmap, which guarantees page aligned allocations, which is closer to the behavior the caller would have gotten specifying HUGEPAGES.  A user can still set MMAP without HUGEPAGES.

Comment (nzmsv) [reply include/ofi_mem.h:292]: I understand the argument, but I think this will lead to brittle code, and I am not sure that the benefits outweigh this cost.  Testing for HUGEPAGES would now need to also be aware of the special case.  Instead of 
```
if (pool->attr.flags & OFI_BUFPOOL_HUGEPAGES) {
}
if (pool->attr.flags & OFI_BUFPOOL_MMAP) {
}
```
one would now have to write:
```
if (pool->attr.flags & (OFI_BUFPOOL_HUGEPAGES & ~OFI_BUFPOOL_MMAP)) {
}
if (pool->attr.flags & OFI_BUFPOOL_MMAP) {
}
```
otherwise `flags == OFI_BUFPOOL_MMAP` would match the first `if` statement (that expression is non-zero).

While I can do this for this review, I fear that this is just creating a footgun for some future maintainer not aware of this conversation.

That said, I am open to making the change.  If the decision is to combine the bits it would probably be cleaner to provide some predicate functions like `bufpool_uses_hugepages` and `bufpool_uses_mmap` to encapsulate the ugliness above.

Comment (shefty) [reply include/ofi_mem.h:292]: Hmm... that's a good point.  The real objective from the viewpoint of the user is to obtain page aligned allocations.  Maybe the implementation stays as proposed, and we really just need a better name for the flag.  It really doesn't matter if we use mmap (which is the easy implementation) or a more complex algorithm with malloc.  The page alignment must hold true, and it can't safely fallback to a standard malloc without silently causing issues (memory corruption in this case).  The hugepage flag is basically defined as a preference (and maybe should have been called HUGEPAGES_PREFERRED). That makes it clear that non-huge pages are acceptable.  But this flag is really forcing a requirement that must be maintained (PAGE_ALIGNED?).  Keeping the flag still keeps the patch simple without going to the extreme of trying to support a generic alloc_alignment.

Comment (nzmsv) [reply include/ofi_mem.h:292]: Thinking about this more, the requirement is actually `mmap` rather than page alignment.  Imagine that we use `posix_memalign` to obtain a page-aligned buffer and called `ibv_reg_mr` on it, setting `MADV_DONTFORK` as a side effect.  Then we fork, and in the child these pages are not mapped.  Even if the child can successfully call `free()` on this chunk, it'll just explode the next time malloc tries to reuse this memory with a hole in it.

In summary: it is never safe to use malloc for these allocations and they must come from a separate mapping, so I think we should call this flag MMAP.

**Review by rashikakheria — APPROVED** [review]

File: include/ofi_mem.h (hunk 9 above)
Comment (wzamazon) [reply include/ofi_mem.h:292]: @nzmsv I think what you said make sense. The main concern I have is the fallback behavior for huge page.

Because huge page implies using mmap, when huge page allocation failed, it should fall back to  mmap.

I suggest to add the  OFI_BUFPOOL_MMAP flag when huge page allocation fail.

Comment (nzmsv) [reply include/ofi_mem.h:292]: @wzamazon I am adding the flag here: https://github.com/nzmsv/libfabric/pull/1

Not adding the flag by default gives more flexibility to the caller (they can choose to fall back to mmap or small pages) but I suppose it's another chance to make a mistake.  I'm OK with either solution.

Comment (nzmsv) [reply include/ofi_mem.h:292]: I added the flag to the fallback from hugepages to mmap.  Currently we'll only fall back from mmap to malloc if the error returned is `FI_ENOSYS` (so we are on an OS that does not support mmap).

**Comment by shefty** [comment]
I've lost the plot here because I don't even know if the problem is being stated correctly.  This is my understanding of the problem:

Memory allocated for a message buffer was registered.  This resulted in the physical page(s) being marked with DONTFORK.  However, the message buffer ended up sharing the same physical page as another data structure which was needed by the forked child.  This resulted in a memory corruption.

*If* this is correct, then the solution to this problem is to ensure that message buffers do not share physical pages with anything that might be needed by a child.  Aligning the memory allocation of the message buffers to a page boundary, _plus_ having the allocation be a multiple of a page size would ensure this.  (If we don't allocate a multiple of page size, then we need some way to prevent the last page of the allocation from being shared.  I don't see that mmap guarantees that.)

Using mmap for the allocation is a convenient way to align the memory allocation of the message buffers to a page boundary.  But I don't see that using mmap is a requirement at all.  It looks to me that the user of the buffer pool wants a guarantee that the buffers cannot share physical pages with anything that isn't part of the pool.  E.g. OFI_BUFPOOL_NONSHARED.  That goes back to having an aligned allocation and an allocation sized to avoid sharing.

File: prov/util/src/util_buf.c [hunk 10]
```diff
@@ -45,6 +45,74 @@ enum {
...
+				buf_region->flags = OFI_BUFPOOL_HUGEPAGES | OFI_BUFPOOL_MMAP;
+				pool->alloc_size = alloc_size;
+				pool->region_size = pool->alloc_size - pool->entry_size;
+				return 0;
+			}
+		}
+		/* If we can't allocate huge pages, fall back to normal
+		 * allocations for all future attempts.
+		 */
+		pool->attr.flags &= (~OFI_BUFPOOL_HUGEPAGES) | OFI_BUFPOOL_MMAP;
```
Comment (wzamazon) [inline prov/util/src/util_buf.c:71]: Is this correct? I think `OFI_BUFPOOL_MMAP` need to be added to flag. so should be `|=`

File: prov/util/src/util_buf.c [hunk 11]
```diff
@@ -45,6 +45,74 @@ enum {
...
+			ret = ofi_alloc_hugepage_buf((void **) &buf_region->alloc_region,
+					     alloc_size);
+			if (!ret) {
+				buf_region->flags = OFI_BUFPOOL_HUGEPAGES | OFI_BUFPOOL_MMAP;
+				pool->alloc_size = alloc_size;
+				pool->region_size = pool->alloc_size - pool->entry_size;
+				return 0;
+			}
+		}
+		/* If we can't allocate huge pages, fall back to normal
```
Comment (wzamazon) [inline prov/util/src/util_buf.c:68]: The comment should be `fallback to mmap allocation`

**Review by wzamazon — CHANGES_REQUESTED** [review]

**Comment by nzmsv** [comment]
> Memory allocated for a message buffer was registered. This resulted in the physical page(s) being marked with DONTFORK. However, the message buffer ended up sharing the same physical page as another data structure which was needed by the forked child. This resulted in a memory corruption.

Yes, that's the current failure that we have seen, however there is another potential issue...

> _If_ this is correct, then the solution to this problem is to ensure that message buffers do not share physical pages with anything that might be needed by a child. 

I think that's the crux of the issue: what is needed by the child?  And the answer to that depends on whether we consider libfabric fork-safe or not.

As written, neither the `mmap` nor the `malloc` solution result in a fork-safe libfabric (and by "safe" I mean "usable in the child after fork()", so perhaps there's a better word than "safe" for this).  Either solution does give us a safe libfabric in the sense that it doesn't crash as long as the child never touches it.

> If we don't allocate a multiple of page size, then we need some way to prevent the last page of the allocation from being shared. I don't see that mmap guarantees that.

Well, `mmap` guarantees it implicitly: any other mapping would have to start on a page boundary simply because there's no way to specify anything else in a page table.  There's no way to create another mapping that will share a chunk of that last page.  This also means that `mmap` is incapable of allocating anything that is not a multiple of page size.

> Using mmap for the allocation is a convenient way to align the memory allocation of the message buffers to a page boundary. But I don't see that using mmap is a requirement at all. 

Yes, for the original failure that we encountered in the field.  However, using `mmap` rather than `malloc` gets us one step closer to a libfabric that is usable after `fork()` (if we decide that is a valid goal).  Why?  Imagine what it would take to reinitialize libfabric in the child.  Presumably it would deallocate this message buffer and allocate it again.  Unfortunately with `malloc` this is guaranteed to go kaboom even if the allocation is aligned.  `mmap` would still work.

But this is all hypothetical because there's likely a lot more work that needs to be done before libfabric can be used after fork in the child process.

File: prov/util/src/util_buf.c (hunk 10 above)
Comment (nzmsv) [reply prov/util/src/util_buf.c]: Yes of course!

**Comment by shefty** [comment]
libfabric does not have a goal of being usable across fork, at least to the extend that any fabric objects may be shared with a child.  (Even MPI doesn't support fork in this manner).  That's unlikely to happen in any foreseeable future.

To move this forward, please update the commit message to describe the exact problem in at least the same amount of detail as above.  I would also rename the proposed MMAP flag to NONSHARED, or something that describes the desired behavior, not implementation.  That would address my earlier concern that huge pages are also allocated using mmap, but keep the hugepage flag independent from this flag.

**Review by wzamazon — APPROVED** [review]

**Review by shefty — APPROVED** [review]
Thank!  This looks good.

**Comment by wzamazon** [comment]
bot:aws:retest

**Comment by wzamazon** [comment]
For some reason, AWS CI was not run on this PR. I just started.

**Comment by wzamazon** [comment]
All tests have passed, and PR has been approved. merging ...
